Add dcrtime page

s-ben opened the pull request:
This PR adds a page for dcrtime under /advanced ( Closes #955 ).

I think this draft is solid, but have left a few questions {in brackets} for technical reviewers.

marcopeereboom commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+1. A 32-byte hash is created for each document or piece of data to be timestamped.
+2. A single merkle root is calculated from the hashes.
+3. This merkle root is included in an on-chain transaction, "anchoring" the data in the Decred blockchain by timestamping the data.
+4. A transaction hash and merkle path is generated, which can then be used to query the Decred blockchain and retreive the block containing the transaction (and associated timestamp).
```
retrieve

s-ben replied:
Fixed

marcopeereboom commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+`8496855341883fdc90cc532f8304d1c46a60586fb15d99f07e41bb5ab19c79c6`
+
+It also creates a "Collection timestamp" corresponding to the time the file was received by the server {SB: true?}.
```
This is correct.

marcopeereboom commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+`Collection timestamp: 1497009600`
+
+If we immediately ask the server if the timestamp has been verified, it will show `Not anchored`. This is because our transaction containing the digest (specifically the merkle root of the digest) has not yet been mined into a block.
```
Merkle root has nothing to do with this. You'll get `Not anchored` until it is time to generate the Merkle which then is anchored. This happens around every start of the hour.

s-ben replied:
Changed to this:

>If we immediately ask the server if the timestamp has been verified, it will show `Not anchored`. This is because our transaction containing the digest has not yet been mined into a block.

marcopeereboom commented on docs/advanced/dcrtime.md:
````diff
@@ -0,0 +1,105 @@
+cd90cc268d9ceef6e276bfa7a615c5f85b5a27b0d917ee3f1f1e5d5598f2fa00 Not anchored
+```
+
+If we wait until the transaction containing the merkle root has been mined into a block and call dcrtime again, the dcrtime server will verify that the transaction has been anchored, as well as provide the merkle path (path to the merkle root) {SB: accurate?}, the transaction ID (TxID) of the transaction containing the merkle root, and a `Chain Timestamp` providing the time the block was mined. 
````
Correct.

marcopeereboom commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+Note that the dcrtime client does verify the merkle path that the server returns. 
+
+Our file is now anchored in the Decred blockchain. Any third party can now prove that `myfile.txt` existed, in its exact form, at the time the block was mined (Chain Timestamp) {SB: is the 'Cellection timestamp relevant here?'}. This can be proved by taking our original `myfile.txt` file, calculating its merkle root, and comparing it to the merkle root stored in the transaction submitted by dcrti ...
```
Collection timestamp seize to be relevant once we have a chain timestamp but since we collect the data we show it. It can be used with some measure of certainty that the server saw the hash prior to the anchoring process. May or may not have value.

jholdstock replied:
> This can be proved by taking our original `myfile.txt` file, calculating its merkle root

Does this make sense? The process for creating a merkle root is described above - take a collection of hashes and concat+hash them down until you only have one hash left.

How can you "generate the merkle root" of a single file? And how would that correspond to the merkle root which has been anchored?

s-ben replied:
Don't mention Chain Timestamp.

marcopeereboom replied:
I don't follow that at all. We calculate the merkle root by sorting the digests and then feeding them into the algorithm.

s-ben replied:
So I wasn't exactly sure how this works, but my guess is based on looking at how [politeia_verify](https://github.com/decred/politeia/tree/master/politeiad/cmd/politeia_verify) works. Iirc, it verifies proposal censorship by creating the digest from the original source file (or concatenated set of files), calculating the merkle root from that, and comparing it to the merkle root in the censorhsip record. To fully verify this against the actual blockchain, I assume one would need to somehow find that merkle root in the blockchain. 

We can also just remove this text and simply point people to the [Politeia Censorship](https://docs.decred.org/governance/politeia/politeia-censorship/#proving-censorship) page, which points people to the politeia_verify tool.

s-ben replied:
Revisiting this, I don't think we need this text actually, as we explain the process already in the 'Timestamping process' section. Have removed.

marcopeereboom commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+Building on our simple example, the following example illustrates how dcrtime is used to timestamp data in Politeia.
+
+Each hour, the Politeia server commits all data uploaded to it (proposals, votes, comments, etc.) to a public git [repo](https://github.com/decred-proposals/mainnet). It then creates {SB: sequence correct?} a hash of each commit, timestamps each commit hash using dcrtime, and anchors the hashes on the Decred blockchain. Once the transaction containing the merkle root of the hash is mined, the cor ...
```
This is incorrect. Pi commits into git as needed (edits, new props etc) and before anchoring it flushes all comments and vote journals. once git has stabilized the last commit is anchored + all other commit hashes up to the prior anchor.

s-ben replied:
Went back and rewrote this section to reflect above, as well as fleshing out the example.

marcopeereboom replied:
Each git commit has an implict hash. We anchor those hashes.

s-ben replied:
Not sure exactly what an implicit hash is, even after a few minutes of googling...Have changed the text to this. Does this work? 

>...These journal files are then committed to the repo when other data such as a new proposal is submitted, or every hour if no other data is submitted. Once Politeia verifies that these commits are in the repo, dcrtime is used to anchor the implicit hash for each commit. For each hash anchored, an anchor file is committed to the `/anchors` directory in the repo. To fasciliate verification, metadata for each anchor is also appended to a single text file, `anchor_audit_trail.txt`, which provides a chonological record of anchors.

marcopeereboom commented on docs/advanced/navigating-politeia-data.md:
```diff
@@ -12,7 +12,7 @@ Below is a screenshot of the root directory of the [decred-proposals/mainnet](ht
 ![The root of the mainnet repository. * = proposal folders, A = anchoring data](/img/politeia/mainnet-pi-repo.png)
 
-Folders marked with `*` contain proposal data. Folders and files marked with `A` contain anchoring data. Anchoring data relates to the anchoring of Politeia data to the Decred blockchain through [dcrtime](https://github.com/decred/dcrtime), which facilitates Politeia's transparent censorship functionality. This guide will not describe those files in detail, as they are not likely to be of interes ...
+Folders marked with `*` contain proposal data. Folders and files marked with `A` contain anchoring data. Anchoring data relates to the anchoring of Politeia data to the Decred blockchain through [dcrtime](dcrtime.md), which facilitates Politeia's transparent censorship functionality. This guide will not describe those files in detail, as they are not likely to be of interest for analysis. 
```
Can't see the png so make sure it matches the comments I made earlier.

s-ben replied:
Both images in the dcrtime page were taken from the original [dcrtime blog](https://blog.decred.org/2017/06/14/Dcrtime/). I removed the Decred logo from both images, but otherwise kept them the same. Below are screenshots of both images.

![image](https://user-images.githubusercontent.com/1634777/60697434-1fae0e00-9e9f-11e9-925b-809b66641652.png)

![image](https://user-images.githubusercontent.com/1634777/60697443-29d00c80-9e9f-11e9-8102-c0d449be4c16.png)

marcopeereboom commented on docs/governance/politeia/overview.md:
```diff
@@ -25,7 +25,7 @@ There is a fee for submitting a proposal (0.1 DCR) to limit the potential for pr
 When proposals are submitted, they are checked by Politeia administrators. Proposals that are deemed spam or invalid will be censored.
 
-Politeia is built around the concept of **transparent censorship**, using [dcrtime](https://github.com/decred/dcrtime). Users cannot be silently censored; they can prove that censorship has occurred. When a user registers, a **cryptographic identity** (pub/priv key pair) is created. This cryptographic identity is then used to create a "censorship record" for each user submission (proposal, commen ...
+Politeia is built around the concept of **transparent censorship**, using [dcrtime](../../advanced/dcrtime.md). Users cannot be silently censored; they can prove that censorship has occurred. When a user registers, a **cryptographic identity** (pub/priv key pair) is created. This cryptographic identity is then used to create a "censorship record" for each user submission (proposal, comment, comme ...
```
`censorship tokens`

s-ben replied:
Fixed

jholdstock commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+---
+
+Dcrtime is a timestamping application that allows users to store hashes of arbitrary data on the Decred blockchain. Inspired by Peter Todd's [opentimestamps](https://petertodd.org/2016/opentimestamps-announcement), dcrtime was originally built to store timestamped hashes of data from [Politeia](../../governance/politeia/overview), Decred's proposal system, on the Decred blockchain. 
```
OpenTimestamps caps.

jholdstock replied:
Link to `overview.md` file

s-ben replied:
Done

jholdstock replied:
answering your question below, the correct path for the link should be `../governance/politeia/overview.md`

s-ben replied:
Fixed

jholdstock commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+Dcrtime is a timestamping application that allows users to store hashes of arbitrary data on the Decred blockchain. Inspired by Peter Todd's [opentimestamps](https://petertodd.org/2016/opentimestamps-announcement), dcrtime was originally built to store timestamped hashes of data from [Politeia](../../governance/politeia/overview), Decred's proposal system, on the Decred blockchain. 
+
+Although dcrtime was created for Politeia, it is expected to have generic utility as a timestamping service. Dcrtime is implemented as a client and server application, with all code open source and  available in the [dcrtime repo](https://github.com/decred/dcrtime). Decred also offers a free timestamping service through its public mainnet server. We expect this to be particularly useful in scenar ...
```
double space "and  available"

jholdstock replied:
Could link "public mainnet server" to https://timestamp.decred.org/ ?

s-ben replied:
Good catch. Good idea. Done

jholdstock commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,105 @@
+
+Dcrtime allows a nearly unlimited number of hashes to be timestamped. It does this by creating a single merkle root from any number of hashes and including it in a transaction on the Decred blockchain. This process can be summarized as follows:
+
+1. A 32-byte hash is created for each document or piece of data to be timestamped.
```
Use `1.` for all items in a numbered list please

s-ben replied:
Ah, keep forgetting. Done

s-ben commented:
@marcopeereboom have addressed comments. Addressing your comment on the Politeia example made me investigate more how it worked. Ended up rewriting most of the exmample and expanding it a  bit. I think it's solid, but definitely needs a sanity check. Also open to just simplifying this back to a single paragraph.

marcopeereboom commented:
I did a once over and there are a few items remaining. Hope the comments made sense.

s-ben commented:
@marcopeereboom @jholdstock have addressed all comments, except the open question of how to describe the process for verifying timestamped files against the blockchain. Still not quite sure how that works. Also my rewording around 'implicit hashes' may need tweaking.

s-ben commented:
P.S. @jholdstock adding `.md` to the Politeia overview page path is causing the link to break. Have changed it back to,
`[Politeia](../../governance/politeia/overview)`, and it's working again. What is the correct syntax here?

s-ben commented:
@marcopeereboom upon reviewing this again, realized we don't really need the text that is holding this PR up. Have removed. That leaves only my rewording around implicit hashes. If that looks good, this page should be good to go.

s-ben commented:
@marcopeereboom for the simple dcrtime example, I just copied the example outputs from your dcrtimed [blog](https://blog.decred.org/2017/06/14/Dcrtime/), as I don't have dcrtimed set up locally. If you can provide sample dcrtime_checker output I can include.

Likewise with Pi, I've taken the example there about as far as my current understanding could take me...which is just examining an [anchor commit](https://github.com/decred-proposals/mainnet/commit/233708a380061982fd44bfb64a5ace325cf59315) for the bug bounty proposal and then looking up the transaction and associated merkle root in the block explorer. Would verification require another step?

s-ben commented:
Tweaked sentence around using dcrtime as a generic timestamping service to make it more straightforward, concise, and less "marketingy".

s-ben commented:
Adding link to dcrtimegui

jholdstock commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,121 @@
+
+Building on our simple example, the following example illustrates how dcrtime is used to timestamp data in Politeia.
+
+Politeia commits user data (new proposals, edits, etc.) as needed to a public git [repo](https://github.com/decred-proposals/mainnet). For performance reasons, incoming vote and comment data are stored in the Politeia backend in "journal" files. These journal files are then committed to the repo when other data such as a new proposal is submitted, or every hour if no other data is submitted. Once ...
```
facilitate*

s-ben replied:
Fixed

jholdstock commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,121 @@
+
+![dcrtime politeia](/img/dcrtime_politeia_diagram.png)
+
+To illustrate, let's examine an anchor from a real proposal, the [Decred Bug Bounty Proposal: Phase 2](https://proposals.decred.org/proposals/073694ed82d34b2bfff51e35220e8052ad4060899b23bc25791a9383375cae70) proposal. This anchor hashes two commits to the blockchain. The first [commit](https://github.com/decred-proposals/mainnet/commit/9125d351db4a429681cd7158d2c17d62a2b47c4c) contains all vote d ...
```
The page uses a mix of "Anchor" and "anchor". Should probably standardize

s-ben replied:
Went with lowercase anchor.

s-ben commented on docs/advanced/dcrtime.md:
```diff
@@ -0,0 +1,121 @@
+# <img class="dcr-icon" src="/img/dcr-icons/Servers.svg" /> Dcrtime
```
Fixed

s-ben commented on mkdocs.yml:
```diff
@@ -116,6 +116,7 @@ nav:
   - 'Block Header Specifications': 'advanced/block-header-specifications.md'
   - 'Blockchain Parameters': 'advanced/blockchain-parameters.md'
   - 'Transaction Details': 'advanced/transaction-details.md'
+  - 'Dcrtime': 'advanced/dcrtime.md'
```
Fixed

s-ben commented:
@jholdstock feedback addressed. Merge moon?
